src/mbgl/renderer/layers/render_location_component_layer.cpp

```diff
@@ -0,0 +1,895 @@
+                                           const TransformState& s,
+                                           LatLng::WrapMode wrapMode = LatLng::Wrapped) {
+        ScreenCoordinate flippedPoint = p;
+        flippedPoint.y = s.getSize().height - flippedPoint.y;
+        return s.screenCoordinateToLatLng(flippedPoint, wrapMode);
+    }
+
+    void setTextureFromFile(const std::string& imagePath,
+                            std::shared_ptr<Texture>& tex,
+                            const mbgl::style::LocationComponentRenderParameters& params) {
+        if (textures.find(imagePath) == textures.end()) {
+            std::shared_ptr<Texture> tx = std::make_shared<Texture>();
+            if (!imagePath.empty() && params.imageManager)
+                tx->assign(&(params.imageManager->getImage(imagePath)->image));
+            else
+                tx->assign(nullptr);
+            textures[imagePath] = tx;
+        }
+        tex = textures.at(imagePath);
+    }
+
+    std::map<std::string, std::shared_ptr<Texture>> textures;
+    mapbox::cheap_ruler::CheapRuler ruler;
+    SimpleShader simpleShader;
+    TexturedShader texturedShader;
+    Buffer buffer;
+    Buffer circleBuffer;
+    Buffer shadowBuffer;
+    Buffer puckBuffer;
+    Buffer hatBuffer;
+    Buffer texCoordsBuffer;
+    std::shared_ptr<Texture> texShadow;
+    std::shared_ptr<Texture> texPuck;
+    std::shared_ptr<Texture> texPuckHat;
+
+    std::array<vec2, 73> circle;        // 72 points + position
+    std::array<vec2, 4> shadowGeometry; // drawing 3 quads using triangles.
+    std::array<vec2, 4> puckGeometry;
+    std::array<vec2, 4> hatGeometry;
+    std::array<vec2, 4> texCoords;
+    mbgl::mat4 projectionCircle;
+    mbgl::mat4 projectionPuck;
+
+    bool positionChanged = false;
+    bool radiusChanged = false;
+    bool bearingChanged = false;
+    mbgl::style::LocationComponentRenderParameters oldParams;
+
+public:
+    mbgl::style::LocationComponentRenderParameters parameters;
+};
+} // namespace style
+
+using namespace style;
+namespace {
+
+inline const LocationComponentLayer::Impl& impl(const Immutable<style::Layer::Impl>& impl) {
+    assert(impl->getTypeInfo() == LocationComponentLayer::Impl::staticTypeInfo());
+    return static_cast<const LocationComponentLayer::Impl&>(*impl);
+}
+
+} // namespace
+
+RenderLocationComponentLayer::RenderLocationComponentLayer(Immutable<style::LocationComponentLayer::Impl> _impl)
+    : RenderLayer(makeMutable<LocationComponentLayerProperties>(std::move(_impl))),
+      host(new PuckLayerHost()),
+      unevaluated(impl(baseImpl).paint.untransitioned()) {
+    assert(gfx::BackendScope::exists());
+    MBGL_CHECK_ERROR(host->initialize());
+}
+
+RenderLocationComponentLayer::~RenderLocationComponentLayer() {
+    assert(gfx::BackendScope::exists());
+    if (contextDestroyed) {
+        host->contextLost();
+    } else {
+        MBGL_CHECK_ERROR(host->deinitialize());
+    }
```

Comment:
this is not a custom layer, not sure why it is here.

Reply:
since this layer is currently drawn using native gl calls, not going through mbgl Programs, buckets, etc, the resource release has to be done